### ISSUE TITLE
Crash in service.c

### ISSUE THREAD

**Issue by ondraco** [warmcat/libwebsockets#441]
Hello,

I have found a new issue which i believe is continuation to #437.

It is now possible to connect without a problem but after refreshing page several times the server stops responding and then crashes. This happens always on what seems to be a fixed number of refreshes (3 in my case refreshes and about 23 files delivered totaly).

The crash happens in `service.c` line 324 
`while (*pwsi) {` anw pwsi seems to be already deleted.

Log right before the crash:
```
lws_read: read_ok, used 381
Accepted 02F014B0 to tsi 0
lws_header_table_attach: wsi 02F014B0: ah 00000000 (tsi 0)
lws_header_table_attach: adding 02F014B0 to ah waiting list
Accepted 062E5950 to tsi 0
Accepted 062E5D58 to tsi 0
Accepted 062E62B8 to tsi 0
Accepted 02EC43C8 to tsi 0
Accepted 02EC4550 to tsi 0
lws_header_table_attach: wsi 062E5950: ah 00000000 (tsi 0)
lws_header_table_attach: adding 062E5950 to ah waiting list
lws_header_table_attach: wsi 062E5D58: ah 00000000 (tsi 0)
lws_header_table_attach: adding 062E5D58 to ah waiting list
lws_header_table_attach: wsi 062E62B8: ah 00000000 (tsi 0)
lws_header_table_attach: adding 062E62B8 to ah waiting list
lws_header_table_attach: wsi 02EC43C8: ah 00000000 (tsi 0)
lws_header_table_attach: adding 02EC43C8 to ah waiting list
lws_header_table_attach: wsi 02EC4550: ah 00000000 (tsi 0)
lws_header_table_attach: adding 02EC4550 to ah waiting list
lws_service_timeout_check: pwsi=02EC4554
lws_service_timeout_check: pwsi=02EC43CC
lws_service_timeout_check: pwsi=062E62BC
lws_service_timeout_check: pwsi=062E5D5C
lws_service_timeout_check: pwsi=062E5954
lws_service_timeout_check: pwsi=02F014B4
*** not on ah wait list ***
lws_close_free_wsi: real just_kill_connection: 062E5B80
remove_wsi_socket_from_fds: wsi=062E5B80, sock=1144, fds pos=17, end guy pos=25, endfd=0
lws_service_timeout_check: pwsi=02EC4554
lws_service_timeout_check: pwsi=02EC43CC
lws_service_timeout_check: pwsi=062E62BC
lws_service_timeout_check: pwsi=062E5D5C
lws_service_timeout_check: pwsi=062E5954
lws_service_timeout_check: pwsi=02F014B4
*** not on ah wait list ***
lws_close_free_wsi: real just_kill_connection: 02EC5418
remove_wsi_socket_from_fds: wsi=02EC5418, sock=1172, fds pos=14, end guy pos=24, endfd=1196
lws_close_free_wsi: real just_kill_connection: 02EC4550
remove_wsi_socket_from_fds: wsi=02EC4550, sock=1196, fds pos=17, end guy pos=23, endfd=1188
lws_service_timeout_check: pwsi=02EC4554
lws_service_timeout_check: pwsi=FEEEFEF2
```

Stack:
```
lws_service_timeout_check(lws * wsi=0x02ec43c8, unsigned int sec=1456397611)  Line 324 + 0x3 bytes	C
lws_service_fd_tsi(lws_context * context=0x065f90b8, lws_pollfd * pollfd=0x00000000, int tsi=0)  Line 548 + 0xd bytes	C
lws_service_fd(lws_context * context=0x065f90b8, lws_pollfd * pollfd=0x00000000)  Line 861 + 0xf bytes	C
lws_plat_service_tsi(lws_context * context=0x065f90b8, int timeout_ms=0, int tsi=0)  Line 206 + 0xb bytes	C
lws_plat_service(lws_context * context=0x065f90b8, int timeout_ms=0)  Line 263 + 0xf bytes	C
lws_service(lws_context * context=0x065f90b8, int timeout_ms=0)  Line 899 + 0xd bytes	C
```

Thank you for your help!

**Comment by lws-team** [comment]
I think I found it just from your log... please try current v1.7-stable

**Comment by ondraco** [comment]
Thanks it crashes no more but still happens that after several refreshes the server stops working. It continues to do the service loop but wont get any new connections. It is possible that it is just my mistake  somewhere I will try to investigate it some more.

**Comment by lws-team** [comment]
I can't reproduce that on the test server, I refreshed it 20x no problem and even running ab against it with 10,000 connections works fine.

It might be worth turning up the logging and seeing what it logs when you try to connect after the problem appears.  Also what is info.max_http_header_pool set to at context creation time?  Setting it to 0 is legal it defaults to 16 ah in the pool then.

**Comment by ondraco** [comment]
It actually stops writing any logs (but the according to netstat the port is still listening). I had the info.max_http_header_pool = 0 so it defaulted to 16.I have tried setting it 100 but there is no effect. I have one context that has two protocols: one for http and one for WS and second context that has only WS (the http is the only one that is dying).

The last logs from last two file requests:
```
DEBUG: Client requested file:/scf.js
DEBUG: Sending file:scf.js
lws_set_timeout: 03058558: 20 secs
_lws_rx_flow_control: no pending change
lws_http_transaction_completed: wsi 03058558
lws_set_timeout: 03058558: 0 secs
lws_header_table_detach: wsi 03058558: ah 06802AC0 (tsi=0, count = 1)
lws_http_transaction_completed: 03058558: keep-alive await new transaction
DEBUG: TOTAL FILES SERVED: 69
lws_read: thinks we have used 0
lws_read: read_ok, used 326
fd=1524, revents=1
lws_header_table_attach: wsi 030586E8: ah 00000000 (tsi 0)
lws_header_table_attach: wsi 030586E8: ah 06802AC0: count 1 (on exit)
lws_server_socket_service: 030586E8: rxpos:0 rxlen:0
lws_server_socket_service: wsi 030586E8, ah->rxlen = 322
lws_read: incoming len 322
lws_set_timeout: 030586E8: 0 secs
No upgrade
lws_union_transition: 030586E8: mode 1
lws_handshake_server: wsi 030586E8: ah 06802AC0
lws_ensure_user_space: 030586E8 protocol 00EBBC68
lws_ensure_user_space: 030586E8 protocol pss 0, user_space=0
Method: GET request for '/script.js'
lws_set_timeout: 030586E8: 20 secs
DEBUG: Client requested file:/script.js...
DEBUG: Sending file:/script.js...
lws_set_timeout: 030586E8: 20 secs
lws_set_timeout: 030586E8: 20 secs
lws_set_timeout: 030586E8: 20 secs
lws_set_timeout: 030586E8: 20 secs
lws_set_timeout: 030586E8: 20 secs
lws_set_timeout: 030586E8: 20 secs
lws_set_timeout: 030586E8: 20 secs
lws_set_timeout: 030586E8: 20 secs
lws_set_timeout: 030586E8: 20 secs
lws_set_timeout: 030586E8: 20 secs
lws_set_timeout: 030586E8: 20 secs
lws_set_timeout: 030586E8: 20 secs
lws_set_timeout: 030586E8: 20 secs
lws_set_timeout: 030586E8: 20 secs
lws_set_timeout: 030586E8: 20 secs
lws_set_timeout: 030586E8: 20 secs
_lws_rx_flow_control: no pending change
lws_http_transaction_completed: wsi 030586E8
lws_set_timeout: 030586E8: 0 secs
lws_header_table_detach: wsi 030586E8: ah 06802AC0 (tsi=0, count = 1)
lws_http_transaction_completed: 030586E8: keep-alive await new transaction
lws_read: thinks we have used 0
lws_read: read_ok, used 322
fd=916, revents=8
accepted new conn  port 63948 on fd=1532
Accepted 03058878 to tsi 0
lws_adopt_socket: new wsi 03058878
lws_set_timeout: 03058878: 20 secs
insert_wsi_socket_into_fds: 03058878: tsi=0, sock=1532, pos-in-fds=63
```

It always ends with: `insert_wsi_socket_into_fds` and around 70 files served. 

I send files in LWS_CALLBACK_HTTP using:
```
n = lws_serve_http_file(wsi, filePath, mime, 0, 0);
if (n < 0 || ((n > 0) && lws_http_transaction_completed(wsi)))
	return -1; 
```
I create the context simply as:
```
struct lws_context_creation_info info;
memset(&info, 0, sizeof info);
info.port = server->m_port;
info.protocols = m_protocols;
info.gid = -1;
info.uid = -1;
info.user = server;

context = lws_create_context(&info);
```

After this there is just silence...
If you had any idea what I may be wrong (or what may i be doing wrong) i would be grateful.

**Comment by lws-team** [comment]
The only thing I see in the logs is fd=916 is abnormal according to "refresh a page a few times"... default process limit for fds in Fedora anyway is 1024.  It sounds like you a radically leaking file / socket descriptors somewhere (opening and not closing) until no more fds.  That's probably happening outside lws since test apps don't show this and it must be using them up much faster than anything in lws would.

**Comment by ondraco** [comment]
It seems that the sockets opened are not closed. The pos-in-fds (pt->fds_count) is growing up to 60+ from which it stops responding. The browser requests them using `Connection:keep-alive`. For each request a fd is added (10 added) but when the request is done only 1 is removed like this:

```
*** not on ah wait list ***
lws_close_free_wsi: real just_kill_connection: 00FC67E8
remove_wsi_socket_from_fds: wsi=00FC67E8, sock=1100, fds pos=21, end guy pos=22, endfd=0
not calling back closed mode=0 state=0
lws_free_wsi: 00FC67E8, remaining wsi 21
```

Shouldn't there be just one connection created with the `Connection:keep-alive` which is reused for all of the requests?

I use the technique shown in the demo:

```
case LWS_CALLBACK_HTTP:
   n = lws_serve_http_file(wsi, filePath, mime, 0, 0);
   if (n < 0 || ((n > 0) && lws_http_transaction_completed(wsi)))
    	   return -1; /* error or can't reuse connection: close the socket */		

case LWS_CALLBACK_HTTP_FILE_COMPLETION:	
   	 if(lws_http_transaction_completed(wsi))
            return -1;
return 0;
```
It seems that the `lws_http_transaction_completed()` returns always 0 (because of the `Connection:keep-alive`).

I have also tried:
```
case LWS_CALLBACK_HTTP_FILE_COMPLETION:	
            return -1;
```

which causes a timer to close all the connections in 20s. Which is also not good because it may kill possible connection that shall be upgraded and when refreshing fast enough i can still get to the 60+ and freeze the server.

Am I missing something?  

Thanks!

**Comment by lws-team** [comment]
Assuming this is on linux, you could have hundreds or hundres of thousands of fds open without problems depending on the process limit.  60 wouldn't be anything to speak of in itself and should not 'freeze the server'.

The thing I mentioned is the socket descriptor index being unexpectedly high.  Every process gets 3 stdin/out/err as indexes 0 1 and 2.  After that they go up by one with the smallest closed ordinal getting reused if any available.  Each process has his own fd count starting from 0.  Lws naturally uses the fd numbers it's given, like 1100, but he only tracks fds he opened, which you say is 60 or so.  Then who opened the other 1000+ in this process?

That we see socket fd numbers like 1100 flying around in your logs means something else in your process - my guess is your code - if opening fds like crazy and not closing them.  For example, it'd look like this if you had custom logging code that opened the log file each line but never closed it.

Assuming it's linux, run the thing and have a look around in ls -l /proc/PID/fd/* to get an idea what all the fds point to (PID is the process ID of your server, you can find it from ps )

**Comment by ondraco** [comment]
This is on Win32 and it seems that the fd numbers are high from the beginning e.g. even higher than 1000. It seems they are not linearly assigned.

**Comment by lws-team** [comment]
Oh... yes they are opaque handles on win32 without rhyme or reason, it's a red herring then.

In that case does the test server act well for you?

**Comment by lws-team** [comment]
Also the listen socket will listen in netstat until closed, it doesn't imply the program is still sane.  If you break it with a debugger is the any clue from where it was executing?

**Comment by JoakimSoderberg** [comment]
To figure out if there's a leak on windows I'd suggest using http://drmemory.org/ and run your program through it (no special compile flags needed)

**Comment by lws-team** [comment]
I think the leak thing was bogus, though... I was just going by fd numbers.

Make sure you have the latest v1.7-stable as well... something related to things not closing until timed out was fixed yesterday.

**Comment by ondraco** [comment]
Just quickly testing the demo server. I had two small problems which i just quickly and ugly resolved (hope should have no influence on result)

1. snprintf does not exist in VS (not untill 2015..I am stuck on 2008) so I used defs from: [http://stackoverflow.com/questions/2915672/snprintf-and-visual-studio-2010](url)
2. strftime in update status kept throwing asserts then crashed..So i have commented out all protocols except:
```
    {
		"http-only",		/* name */
		callback_http,		/* callback */
		sizeof (struct per_session_data__http),	/* per_session_data_size */
		0,			/* max frame size / rx buffer */
	},
```

Enabled all logging and kept hitting f5 on http://localhost:7681/ untill the `pos-in-fds` gets over 60 then the `lws_service(` returns -1 and server exits.

**Comment by lws-team** [comment]
Hm this doesn't happen here on Fedora, on current v1.7-stable I did F5 on the test server with -d65535 for 120 x.  There's no buildup of open fds.

The final set of closes (there are several open ws links on each page was like this)

remove_wsi_socket_from_fds: wsi=0xf2f490, sock=11, fds pos=1, end guy pos=4, endfd=11

so he recycles the fds every time properly.

**Comment by ondraco** [comment]
So just to be sure I tested on other computer (Win10) using VS2015 so I didn't have to do any changes to make the test-server run (VS2015 has `snprintf() `and `strftime()`supports the %F and %Z which is used in the test server) and to mineralize chance that it is some external influence. The result is almost the same. After few refreshes the server stops responding. It still runs and it still loops the `n = lws_service(context, 50); `.

I have attached the full log :[LOG.txt](https://github.com/warmcat/libwebsockets/files/149348/LOG.txt)

Each subsequent try for connection leads to following output:

```
[2016/01/27 14:38:13:6683] DEBUG: accepted new conn  port 55364 on fd=280
[2016/01/27 14:38:13:6693] INFO: Accepted 027502E8 to tsi 0
[2016/01/27 14:38:13:6703] DEBUG: lws_adopt_socket: new wsi 027502E8
[2016/01/27 14:38:13:6713] DEBUG: insert_wsi_socket_into_fds: 027502E8: tsi=0, sock=280, pos-in-fds=20
[2016/01/27 14:38:13:6723] INFO: lws_header_table_attach: wsi 027502E8: ah 00000000 (tsi 0)
[2016/01/27 14:38:13:6733] INFO: lws_header_table_attach: adding 027502E8 to ah waiting list
[2016/01/27 14:38:13:9185] DEBUG: accepted new conn  port 55365 on fd=712
[2016/01/27 14:38:13:9195] INFO: Accepted 02750440 to tsi 0
[2016/01/27 14:38:13:9205] DEBUG: lws_adopt_socket: new wsi 02750440
[2016/01/27 14:38:13:9265] DEBUG: insert_wsi_socket_into_fds: 02750440: tsi=0, sock=712, pos-in-fds=21
[2016/01/27 14:38:19:0184] NOTICE: wsi 02750440: TIMEDOUT WAITING on 3 (did hdr 0, ah 00000000, wl 1, pfd events 0)
[2016/01/27 14:38:19:0200] INFO: lws_close_free_wsi: real just_kill_connection: 02750440
[2016/01/27 14:38:19:0201] INFO: remove_wsi_socket_from_fds: wsi=02750440, sock=712, fds pos=21, end guy pos=22, endfd=656
[2016/01/27 14:38:19:0221] DEBUG: not calling back closed mode=0 state=0
[2016/01/27 14:38:19:0231] INFO: lws_header_table_detach: wsi 02750440: ah 00000000 (tsi=0, count = 16)
[2016/01/27 14:38:19:0241] DEBUG: lws_free_wsi: 02750440, remaining wsi 21
[2016/01/27 14:38:19:0257] NOTICE: wsi 027502E8: TIMEDOUT WAITING on 3 (did hdr 0, ah 00000000, wl 1, pfd events 0)
[2016/01/27 14:38:19:0268] INFO: lws_close_free_wsi: real just_kill_connection: 027502E8
[2016/01/27 14:38:19:0278] INFO: remove_wsi_socket_from_fds: wsi=027502E8, sock=280, fds pos=20, end guy pos=21, endfd=712
[2016/01/27 14:38:19:0298] DEBUG: not calling back closed mode=0 state=0
[2016/01/27 14:38:19:0308] INFO: lws_header_table_detach: wsi 027502E8: ah 00000000 (tsi=0, count = 16)
[2016/01/27 14:38:19:0328] INFO: lws_header_table_detach: wsi 027502E8, remv wait
[2016/01/27 14:38:19:0348] DEBUG: lws_free_wsi: 027502E8, remaining wsi 20
```

**Comment by lws-team** [comment]
From the log the 'stopping responding' is because the ah pool (default 16) is all used up.  The log shows the new connections not getting an ah in 5s and timing out.

If you change the number of ah in the pool at context creation time, eg

```
	info.max_http_header_pool = 128;
```

it should affect the number of F5s you can survive.

But it doesn't explain why the other connections still have ah attached, they should detatch after the hesders are processed.

**Comment by ondraco** [comment]
I tested it and it sure is connected. The results are:

` info.max_http_header_pool = 16;` -> 8 connections then the server continues to run but does not work

` info.max_http_header_pool = 128;` -> 29 connections then the server turns off completely because of this condition:

```
if (ev < WSA_WAIT_EVENT_0 || ev > WSA_WAIT_EVENT_0 + pt->fds_count)
		return -1;
```

where `pt->fds_count == 64`

Here is the full log for the 128 case: [LOG128.txt](https://github.com/warmcat/libwebsockets/files/149359/LOG128.txt)

**Comment by lws-team** [comment]
From the big log, basically all the http conns are being kept alive, which is arguably reasonable.

However they are not dropping their ah which is wrong.  In lib/server.c, we call

lws_http_transaction_completed(struct lws *wsi)

when we finished serving one set of http headers.  In there, if keepalive, he is supposed to make a decision about the attached ah

```
	if (wsi->u.hdr.ah) {
		if (!wsi->u.hdr.more_rx_waiting) {
			wsi->u.hdr.ah->rxpos = wsi->u.hdr.ah->rxlen;
			lws_header_table_detach(wsi);
		} else
			lws_header_table_reset(wsi);
	}
```

If we exhausted the buffered headers, we detach the ah because no telling when we get more headers from the network.

If there are some pipelined headers already buffered, if keeps the ah and just resets it, because he will get a forced POLLIN service next time around the event loop to deal with the buffered headers and needs the ah for that.

It seems wsi->u.hdr.more_rx_waiting is always set for you, so we always keep and reset the ah.  But in the logs, he never gets more service for the supposedly buffered next headers.  So we keep the ah on an idle connection forever.

Can you please capture with wireshark or so and paste here the packet with the headers for one connection attempt for that browser?  I'm curious if there is something else after the last byte of the first header set.

**Comment by ondraco** [comment]
Here is the capture: [WireSharkOutput.zip](https://github.com/warmcat/libwebsockets/files/149428/WireSharkOutput.zip)

**Comment by lws-team** [comment]
Thanks, that was fine.  There's no pipelining or extra data from the browser.

It seems         if (!wsi->u.hdr.more_rx_waiting)   is just wrong.  I'll study it thismorning.

**Comment by lws-team** [comment]
Both here and in your logs, if the connection upgrades to ws then the ah is correctly detached.  This is from your logs -->

```
[2016/01/27 14:34:16:1620] INFO: Upgrade to ws
[2016/01/27 14:34:16:1620] INFO: checking lws-status
[2016/01/27 14:34:16:1620] INFO: prot match 4
[2016/01/27 14:34:16:1620] INFO: lws_ensure_user_space: 00EA9128 protocol 0103C068
[2016/01/27 14:34:16:1620] PARSER: lws_parse calling handshake_04
[2016/01/27 14:34:16:1620] INFO: lws_ensure_user_space: 00EA9128 protocol 0103C068
[2016/01/27 14:34:16:1620] INFO: lws_ensure_user_space: 00EA9128 protocol pss 808, user_space=15384608
[2016/01/27 14:34:16:1620] PARSER: WSI_TOKEN_EXTENSIONS = 'permessage-deflate; client_max_window_bits'
[2016/01/27 14:34:16:1620] EXTENSION: lws_extension_callback_pm_deflate: LWS_EXT_CB_*CONSTRUCT
[2016/01/27 14:34:16:1620] PARSER: count_act_ext <- 1
[2016/01/27 14:34:16:1620] PARSER: issuing resp pkt 211 len
[2016/01/27 14:34:16:1630] ERR: cache_len 317
[2016/01/27 14:34:16:1630] INFO: lws_handshake_server: 00EA9128: inheriting ah in ws mode (rxpos:0, rxlen:583)
[2016/01/27 14:34:16:1630] DEBUG: lws_union_transition: 00EA9128: mode 3
[2016/01/27 14:34:16:1630] INFO: Allocating RX buffer 144
[2016/01/27 14:34:16:1630] PARSER: accepted v13 connection
[2016/01/27 14:34:16:1630] DEBUG: lws_read: thinks we have used 0
[2016/01/27 14:34:16:1630] INFO: lws_read: read_ok, used 583
[2016/01/27 14:34:16:1630] INFO: lws_header_table_detach: wsi 00EA9128: ah 00814050 (tsi=0, count = 3)
```

But it looks like the browser does something different if it stays in http.  From my logging

```
lwsts[23665]: Accepted 0x16b2f80 to tsi 0
lwsts[23665]: lws_header_table_attach: wsi 0x16b2f80: ah (nil) (tsi 0)
lwsts[23665]: lws_header_table_attach: wsi 0x16b2f80: ah 0x1686150: count 1 (on exit)
lwsts[23665]: No upgrade
lwsts[23665]: lws_ensure_user_space: 0x16b2f80 protocol 0x6295c0
lwsts[23665]: Method: GET request for '/'
    get  = /
    host: = localhost:7681
    connection: = keep-alive
    http/1.1  = HTTP/1.1
    accept: = text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8
    accept-encoding: = gzip, deflate, sdch
    accept-language: = en-US,en;q=0.8
    cache-control: = max-age=0
    cookie: = test=LWS_1456364736_950449_COOKIE
    user-agent: = Mozilla/5.0 (X11; Fedora; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2657.0 Safari/537.36
lwsts[23665]: HTTP connect from localhost.localdomain (127.0.0.1)
lwsts[23665]: test_server_fops_open: opening /usr/local/share/libwebsockets-test-server/test.html, ret 9, len 17703
lwsts[23665]: lws_http_transaction_completed: 0x16b2f80: keep-alive await new transaction
lwsts[23665]: lws_http_transaction_completed: 0x16b2f80: keep-alive await new transaction
lwsts[23665]: lws_read: read_ok, used 451
lwsts[23665]: lws_server_socket_service: read 0 len
lwsts[23665]: lws_close_free_wsi: shutting down connection: 0x16b2f80
lwsts[23665]: lws_server_socket_service: read 0 len
lwsts[23665]: lws_close_free_wsi: real just_kill_connection: 0x16b2f80
lwsts[23665]: remove_wsi_socket_from_fds: wsi=0x16b2f80, sock=8, fds pos=2, end guy pos=3, endfd=0
lwsts[23665]: lws_header_table_detach: wsi 0x16b2f80: ah 0x1686150 (tsi=0, count = 1)
```

He gets the ah, does the headers, serves the file, completes the transaction (he doesn't drop the ah either... I look at that next) and then the browser closes the connection.  He detaches the ah during close processing, hiding the problem he didn't drop the ah during transaction_completed.

From your logs

```
[2016/01/27 14:34:15:7926] INFO: Accepted 00EA7198 to tsi 0
[2016/01/27 14:34:15:7926] DEBUG: lws_adopt_socket: new wsi 00EA7198
[2016/01/27 14:34:15:7926] DEBUG: insert_wsi_socket_into_fds: 00EA7198: tsi=0, sock=468, pos-in-fds=1
[2016/01/27 14:34:15:7926] INFO: lws_header_table_attach: wsi 00EA7198: ah 00000000 (tsi 0)
[2016/01/27 14:34:15:7926] INFO: lws_header_table_attach: wsi 00EA7198: ah 00812750: count 1 (on exit)
[2016/01/27 14:34:15:7926] DEBUG: lws_server_socket_service: 00EA7198: rxpos:0 rxlen:0
[2016/01/27 14:34:15:7926] DEBUG: lws_server_socket_service: wsi 00EA7198, ah->rxlen = 451

[2016/01/27 14:34:15:7926] DEBUG: lws_read: incoming len 451
[2016/01/27 14:34:15:7926] PARSER: issuing 451 bytes to parser
[2016/01/27 14:34:16:0440] PARSER: lws_parse sees parsing complete
[2016/01/27 14:34:16:0440] INFO: No upgrade
[2016/01/27 14:34:16:0440] DEBUG: lws_union_transition: 00EA9680: mode 1
[2016/01/27 14:34:16:0440] DEBUG: lws_handshake_server: wsi 00EA9680: ah 008133D0
[2016/01/27 14:34:16:0440] INFO: lws_ensure_user_space: 00EA9680 protocol 0103C008
[2016/01/27 14:34:16:0440] INFO: Method: GET request for '/libwebsockets.org-logo.png'
[2016/01/27 14:34:16:0440] NOTICE: HTTP connect from OndraBook (192.168.0.100)
[2016/01/27 14:34:16:0450] NOTICE: test_server_fops_open: opening D:/test-server/libwebsockets.org-logo.png, ret 532, len 7029
[2016/01/27 14:34:16:0450] DEBUG: lws_http_transaction_completed: wsi 00EA9680
[2016/01/27 14:34:16:0450] INFO: lws_http_transaction_completed: 00EA9680: keep-alive await new transaction
[2016/01/27 14:34:16:0450] DEBUG: _lws_rx_flow_control: no pending change
[2016/01/27 14:34:16:0450] DEBUG: lws_http_transaction_completed: wsi 00EA9680
[2016/01/27 14:34:16:0450] INFO: lws_http_transaction_completed: 00EA9680: keep-alive await new transaction
[2016/01/27 14:34:16:0450] DEBUG: lws_read: thinks we have used 0
[2016/01/27 14:34:16:0450] INFO: lws_read: read_ok, used 408
```

and 00EA9680 doesn't do anything more until he's closed when you shut down the app.

So that's easier I can reproduce the underlying issue here (don't drop ah when we should) but it was obfuscated by this difference in what seems to be browser connection management (both Chrome).

**Comment by lws-team** [comment]
Well... the problem was the flag used to see if any more headers were already received no longer had the right scope.  The union it was in was no longer valid by the time we checked it, so we read garbage and did the wrong thing.

I just pushed a patch on both master and v1.7-stable branches that fixes it, this should remove the ah pool exhaustion problem completely.  The http connections from your browser seem to stay up, but that's OK and they no longer have an ah attached while they do that.

**Comment by ondraco** [comment]
Nice :) I am not anywhere near computer now but i will try to test it and post the result here later today.

**Comment by ondraco** [comment]
I have tested it and it seems there remains one more problem. It is now possible to refresh at lest 28times but after that the test-server just exits on (`lws_plat_service_tsi()` in` lws-plat-win.c` ):

```
if (ev < WSA_WAIT_EVENT_0 || ev > WSA_WAIT_EVENT_0 + pt->fds_count)
        return -1;
```
where 
`ev = 4294967295` and `pt->fds_count = 64`

LOG: [LOG.txt](https://github.com/warmcat/libwebsockets/files/149913/LOG.txt)

This was happening also before if the `info.max_http_header_pool ` was set hight enough to prevent accepting any more connections.

**Comment by lws-team** [comment]
Good about the ah thing being fixed.

It's returning WSA_WAIT_FAILED.

If you read

https://msdn.microsoft.com/en-us/library/windows/desktop/ms742219(v=vs.85).aspx

You can have a laugh at the (completely ignored) comments.  But

http://forums.codeguru.com/showthread.php?108651-what-is-the-limit-for-WSAWaitForMultipleEvents

...from a long time ago tells the basic issue, that toy api tops out at 64 sockets.

For many tasks that's actually ok (like you only ever have a handful of clients on your server at a time).  Accepting that the limit, the real question is how come the browser (apparently) doesn't close the connection when you do F5, so you reach the limit.  That's not happening here, chrome closes from his side.  Are we maybe not hearing about the client close on windows platform somehow?

**Comment by ondraco** [comment]
Does not seem to be browser related. Same happens on Safari and Edge

**Comment by lws-team** [comment]
I pushed a patch on v1.7-stable to defeat the recently-added extra socket shutdown phase of the close, we can see if the problem relates to that.

If not I'll have to stare at the code and add more logging.

**Comment by ondraco** [comment]
Seem like it is not related. I tested it and it is still the same. Closes because of WSA_WAIT_FAILED. Once I get a bit of time I will also try to look for the specific reason why it is not closing or at least provide more detailed debug information ;-).

**Comment by ondraco** [comment]
To make it easier I have disabled everything except the HTTP protocol and used simple test.html which has just `Hello world` in it (along with all the html headers stuff).

It seems to me that chrome closes the connection also here. It requires two files `index.html` and `favicon.ico`. It sends request for the file (for example `/`) and after it receives it the chrome sends `[FIN, ACK]` It requires second file and sends `[FIN, ACK]` again. But after that they keep exchanging keep-alive for quite some time. Which is correct from chrome's side I guess. By sending the fin he closes his part of connection and actively tells "I wont be sending any data to you any more but if you have any you can send!" And when there is anything waiting on server side it is Ok to close the connection. 

See trace: [Chrome.zip](https://github.com/warmcat/libwebsockets/files/150184/Chrome.zip)

On the other hand I don't believe that this is the core of the problem because for example the same problem happens for Edge. Edge does not request the `favicon.ico`. He requests the `/` and then sends `[FIN, ACK]` but gets no `[FIN]` from the server which still keeps the connection. After some time Edge just sends [RST, ACK] to kill the connection. 

See trace: [Edge.zip](https://github.com/warmcat/libwebsockets/files/150187/Edge.zip)

Where is the place where from where closing should be correctly initiated `lws_service_fd_tsi():close_and_handled`?

**Comment by ondraco** [comment]
I believe i may have found it. Have to test it more to be sure but in: 

`lws_plat_win.c`
`lws_plat_insert_socket_into_fds()`

You specify the network events:

`WSAEventSelect(wsi->sock, pt->events[pt->fds_count], LWS_POLLIN);`

but `LWS_POLLIN `is only `(FD_READ | FD_ACCEPT)` so there is no FD_CLOSE.

So should not it be more like:

` WSAEventSelect(wsi->sock, pt->events[pt->fds_count], LWS_POLLIN | LWS_POLLHUP);`

Then in `WSAEnumNetworkEvents(` you get also the close event.

**Comment by ondraco** [comment]
Ok so first test result on umnodified test-server: was hitting f5 at least 200 times and still working!

**Comment by lws-team** [comment]
Thanks a lot for the debugging.  Basically although we can build-test it via Appveyor, there's nobody who follows development testing with windows on this project.  So after major changes that happened in v1.6 -> v1.7, there is some debugging debt built up.

I pushed your fix on v1.7-stable and master, hopefully that's it.

**Comment by lws-team** [comment]
I'll close this one, please reopen or make a new issue if any more problems coming.
